hammerblade/torch/kernel/hb_elementwise_for.hpp

```diff
@@ -0,0 +1,370 @@
+// =========================================================
+
+template<typename scalar_t, typename F>
+inline void hb_elementwise_for(BSGTensor<scalar_t> res,
+                               BSGTensor<scalar_t> input,
+                               F functor) {
+  char* data[2];
+  data[0] = res.data_ptr();
+  data[1] = input.data_ptr();
+
+  // is_trivial_1d
+  if(res.ndim() == 1) {
+
+    //-----------------------------
+    // collect metadata
+    //-----------------------------
+    uint32_t strides[2];
+    strides[0] = (res.get_strides())[0];
+    strides[1] = (input.get_strides())[0];
+
+    //-----------------------------
+    // iterating over all elementes
+    //-----------------------------
+    size_t start = 0;
+    size_t end = res.numel();
+    for (size_t idx = start; idx < end; idx++) {
+      scalar_t* res_dp = (scalar_t*)(data[0] + strides[0] * idx);
+      scalar_t* input_dp = (scalar_t*)(data[1] + strides[1] * idx);
+      *res_dp = functor(*input_dp);
+    }
+  } else {
+    //-----------------------------
+    // iterating over all elementes
+    //-----------------------------
+    size_t start = 0;
+    size_t end = res.numel();
+    for (size_t idx = start; idx < end; idx++) {
+      scalar_t* res_dp = (scalar_t*)(data[0] + offset_calc(idx, res));
+      scalar_t* input_dp = (scalar_t*)(data[1] + offset_calc(idx, input));
+      *res_dp = functor(*input_dp);
+    }
+  }
+}
+
+// =========================================================
+// Elementwise for -- Nullary ops
+// =========================================================
+
+template<typename scalar_t, typename F>
+inline void hb_elementwise_for(BSGTensor<scalar_t> res,
+                               F functor) {
+  char* data[1];
+  data[0] = res.data_ptr();
+
+  // is_trivial_1d
+  if(res.ndim() == 1) {
+
+    //-----------------------------
+    // collect metadata
+    //-----------------------------
+    uint32_t strides[1];
+    strides[0] = (res.get_strides())[0];
+
+    //-----------------------------
+    // iterating over all elementes
+    //-----------------------------
+    size_t start = 0;
+    size_t end = res.numel();
+    for (size_t idx = start; idx < end; idx++) {
+      scalar_t* res_dp = (scalar_t*)(data[0] + strides[0] * idx);
+      *res_dp = functor();
+    }
+  } else {
+    //-----------------------------
+    // iterating over all elementes
+    //-----------------------------
+    size_t start = 0;
+    size_t end = res.numel();
+    for (size_t idx = start; idx < end; idx++) {
+      scalar_t* res_dp = (scalar_t*)(data[0] + offset_calc(idx, res));
+      *res_dp = functor();
+    }
+  }
+}
+
+// =========================================================
+// Tile Element-wise for -- Binary ops
+//
+// This function calculates the per tile range automatically
+//==========================================================
+
+template<typename scalar_t, typename F>
+inline void hb_tile_elementwise_for(BSGTensor<scalar_t> res,
```

Comment:
What do think about changing `hb_tile_<>_for` to `hb_parallel_<>_for`? Like the `parallel_for` they defined in pytorch...

Reply:
I definitely like hb_parallel_for ... I think we really only need one kind of hb_parallel_for the parallel_for just gives you a linear i and then you can do what you want with that ...

Reply:
ok! I also like the name `hb_parallel_<>_for`